packages/compose/src/utils/create-higher-order-component/index.ts

```diff
@@ -0,0 +1,58 @@
+/**
+ * External dependencies
+ */
+import { camelCase, upperFirst } from 'lodash';
+// eslint-disable-next-line no-restricted-imports
+import type { ComponentType } from 'react';
+import type { Subtract } from 'utility-types';
+
+/**
+ * Higher order components can cause props to be obviated. For example a HOC that
+ * injects i18n props will obviate the need for the i18n props to be passed to the component.
+ *
+ * If a HOC does not obviate the need for any specific props then we default to `{}` which
+ * essentially subtracts 0 from the original props of the passed in component. An example
+ * of this is the `pure` HOC which does not change the API surface of the component but
+ * simply modifies the internals.
+ */
+export interface HigherOrderComponent< TObviatedProps extends object = {} > {
+	< TP extends TObviatedProps >(
+		OriginalComponent: ComponentType< TP >
+	): ComponentType< Subtract< TP, TObviatedProps > >;
+}
+
+/**
+ * Given a function mapping a component to an enhanced component and modifier
+ * name, returns the enhanced component augmented with a generated displayName.
+ *
+ * @param mapComponentToEnhancedComponent Function mapping component to enhanced component.
+ * @param modifierName                    Seed name from which to generated display name.
+ *
+ * @return Component class with generated display name assigned.
+ */
+function createHigherOrderComponent<
+	TObviatedProps extends object,
```

Comment:
Could this be called `TRemovedProps`? I had to look up the word "obviated" 🙂

Reply:
Yup! Removed is was simpler, I just didn't think of it before 😅

Reply:
One more idea: making the naming closer to the programmer's thinking, these props are `TInjectedProps`. The HOC injects them into the inner component. The fact that they are removed from the outer component's type is secondary.